### DISCUSSION TITLE
Tutorial: Run TriliumNext Server in Termux on Android

### DISCUSSION THREAD

**Discussion by Nriver** [TriliumNext/Trilium#5992] [Show and tell]
**Hi, it’s Nate again! :)**

Exactly one year ago, I wrote a tutorial on how to run the original Trilium on Android. You can check it out [here](https://github.com/zadam/trilium/discussions/4542). Today, I'm back with a new guide for TriliumNext. Enjoy!

If you're looking to run TriliumNext on an arm64 pc, you can skip ahead and start directly from Step 5.

---

#### 1. **Install Termux**
  Download and install [Termux](https://termux.dev/en/) from the official website.

#### 2. **Install PRoot Distro**
  Open Termux and install [PRoot Distro](https://github.com/termux/proot-distro), a tool for managing Linux distributions.
  
  PRoot isn’t a virtual machine—learn more about it [here](https://github.com/termux/proot-distro/blob/master/README.md).
  
  Run the following command:
  ```bash
  pkg install proot-distro
  ```

#### 3. **Install a Base System**
  I recommend using Manjaro. You can list available distributions with `proot-distro list`. To install Manjaro:
  ```bash
  proot-distro install manjaro
  ```

#### 4. **Log in to Manjaro**
  Log into the newly installed Manjaro distribution:
  ```bash
  proot-distro login manjaro
  ```

#### 5. **Install Node.js**
  By default, Termux installs the latest version of Node.js, but it's generally better to use a specific, tested version for stability and compatibility. We’ll use `nvm`(Node Version Manager) to install a fixed version.
  
  First, install the necessary packages:
  ```bash
  pacman -S nodejs npm base-devel curl git
  ```

  Then, install `nvm`:
  ```bash
  curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.1/install.sh | bash
  export NVM_DIR="$HOME/.nvm"
  [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
  ```

  Install Node.js, here I use `v20.17.0`:
  ```bash
  nvm install v20.17.0
  nvm use v20.17.0
  node -v
  ```

#### 6. **Get TriliumNext Source Code**
  Unlike the original Trilium, **TriliumNext** cannot be installed via a release file and `npm install`. If you try this, you'll encounter an error, which is explained in the discussion [here](https://github.com/orgs/TriliumNext/discussions/775). Instead, you need to clone the source code directly:
  ```bash
  git clone https://github.com/TriliumNext/Notes trilium
  cd trilium
  ```

#### 7. **Install Global Packages**
  Install the necessary global packages:
  ```bash
  npm install -g nodemon cross-env
  ```

#### 8. **Install Dependencies**
  This step can be tricky. You might encounter errors, so be prepared to retry the command or use a proxy. Run the following:
  ```bash
  npm install
  npm run switch-server
  ```

#### 9. **Start TriliumNext Server**
  Launch the TriliumNext server with the following command:
  ```bash
  TRILIUM_ENV=dev nodemon src/main.ts
  ```

#### 10. **Access TriliumNext**
  - On your phone’s browser: `http://127.0.0.1:8080`
  - On your computer: `http://your_phone_ip:8080`

#### 11. **Quick Startup with a Script**
  To start TriliumNext quickly, create a `.sh` script.

  1. Log into Manjaro:
     ```bash
     proot-distro login manjaro
     ```
  
  2. Create the script:
     ```bash
     vim /root/run-trilium.sh
     ```

  3. Add the following content:
     ```bash
     export NVM_DIR="$HOME/.nvm"
     [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
     nvm use v20.17.0
     cd /root/trilium
     TRILIUM_ENV=dev nodemon src/main.ts
     ```

  4. Save and exit the editor, then make the script executable:
     ```bash
     chmod +x /root/run-trilium.sh
     exit
     ```

  5. Now, to start Trilium quickly, run:
     ```bash
     proot-distro login manjaro -- /bin/sh /root/run-trilium.sh
     ```

---

#### **Using a Proxy**
  If you face internet connectivity issues (e.g., during `curl`, `npm install`, or `git clone`), you can use a proxy with `proxychains`. Here’s how:

  1. Install `proxychains`:
     ```bash
     pacman -S proxychains
     ```

  2. Modify the proxy settings:
     ```bash
     vim /etc/proxychains.conf
     # Example proxy: http 127.0.0.1 1080
     ```

  3. Use `proxychains` to run commands:
     ```bash
     proxychains npm install
     proxychains npm run switch-server
     ```

#### **Clearing Cache**
  If `npm install` keeps failing, try clearing the npm cache and node_modules:
  ```bash
  rm -rf node_modules
  rm -rf ~/.npm/_cacache
  ```

#### **Why is it Slow?**
  If you find TriliumNext is running slowly, try these solutions:
  - **Disable power-saving mode** on your phone.
  - Ensure that **Termux** has permission to run in the background.

**Comment by Nriver** [comment]
![screen.png](https://github.com/user-attachments/assets/24832a81-e7a5-4745-9b0b-002475a7c885)

**Comment by tucoawa** [comment]
Just awesome. Thank you

**Comment by hewoyu** [comment]
# 手机搭建trilium服务器
本人对下面工具，软件只能简单使用，对代码不了解，下面所有的内容出现错误使用AI deepseek解决，很多地方可能出现问题，欢迎来优化，手机使用vivo Y33s，下面更多是对[Nriver 公司](https://github.com/Nriver)的内容补充，大佬就没必要看了，直接看原文章，适合小白
以下是我根据[Nriver 公司](https://github.com/Nriver)的内容重新梳理的TriliumNext Android端部署教程：
原文章https://github.com/orgs/TriliumNext/discussions/827
* * *

在Android设备上通过Termux部署TriliumNext服务端
-----------------------------------

准备工作
----

### 1\. 安装Termux

推荐从以下渠道获取：

*   [F-Droid仓库](https://f-droid.org/packages/com.termux/)
*   [GitHub Releases](https://github.com/termux/termux-app/releases)

### 2\. 配置PRoot环境

```sh

pkg update && pkg upgrade

pkg install proot-distro

```

### 3\. 安装Linux发行版

建议选择Manjaro（兼容性最佳）：

```sh

proot-distro install manjaro


```

结果

```
Error: distribution 'manjaro' is already installed.
Log in:     proot-distro login manjaro
Reinstall:  proot-distro reset manjaro
Uninstall:  proot-distro remove manjaro
```

环境配置
----

### 4\. 进入Linux子系统Manjaro

```sh

proot-distro login manjaro

```

### 5\. 在Manjaro即\[root@localhost ~\]#下安装基础工具

```sh

pacman -Syu  # 更新系统

pacman -S nodejs npm base-devel curl git

```

### 6\. 配置Node版本管理

此处按照原文第六步curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.1/install.sh | bash，国内会出现下面情况
curl: (77) error setting certificate file: /etc/ssl/certs/ca-certificates.crt
解决方法
```
步骤1. 安装证书包
# 在Manjaro子系统内执行

pacman -Sy ca-certificates --noconfirm

步骤2. 更新证书链update-ca-trust force-enable

update-ca-trust extract
```

然后又会出现下面情况
export NVM\_DIR="$HOME/.nvm"  
error: RPC failed; curl 28 Failed to connect to github.com port 443 after 130165 ms: Couldn't connect to server  
fatal: expected flush after ref listing  
Failed to clone nvm repo. Please report this!
建议多试几次，本人也是莫名的就好了
```sh1
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.1/install.sh | bash  
export NVM\_DIR="$HOME/.nvm"  
\[ -s "$NVM\_DIR/nvm.sh" \] && \\. "$NVM\_DIR/nvm.sh"
```
```sh2

curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.1/install.sh | bash

source ~/.bashrc

nvm install v20.17.0

nvm use v20.17.0

```

TriliumNext部署在\[root@localhost ~\]#下进行
--------------------------------------

### 7\. 获取源码

```sh

git clone https://github.com/TriliumNext/Notes trilium

cd trilium
```

可能出现错误：fatal: unable to access 'https://github.com/TriliumNext/Notes/': Failed to connect to github.com port 443 after 129814 ms: Couldn't connect to server

此错误表示git链接出现错误，国内打不开，更换链接如下面的链接

```
git clone https://github.com/TriliumNext/Notes trilium
git clone https://wget.la/https://github.com/TriliumNext/Notes trilium
```

**termux如何删除文件（本人在这里出现了错误需要删除下载，重新下载）**

1.  使用 rm 命令删除文件： rm 文件路径 例如，要删除名为"example.txt"的文件，可以运行以下命令： rm example.txt
    
2.  使用 rm -r 命令删除目录及其内容： rm -r 目录路径 例如
    

### 8\. 安装项目依赖

```sh

npm install -g nodemon cross-env  # 全局工具

npm install  # 项目依赖

npm run switch-server  # 切换服务端模式
```

### 9\. 启动服务

```sh

TRILIUM_ENV=dev nodemon src/main.ts
```
出现下面错误缺少tsx，解决方法，下载pnpm，安装tsx
```
[root@localhost trilium]# TRILIUM_ENV=dev nodemon src/main.ts
[nodemon] 3.1.9
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): src/**/* translations/**/*
[nodemon] watching extensions: ts,js,json
[nodemon] starting `tsx src/main.ts`
sh: line 1: tsx: command not found
[nodemon] failed to start process, "tsx" exec not found
[nodemon] Error
   at Bus.<anonymous> (/usr/lib/node_modules/nodemon/lib/nodemon.js:164:25)
   at Bus.emit (node:events:531:35)
   at ChildProcess.<anonymous> (/usr/lib/node_modules/nodemon/lib/monitor/run.js:199:11)
   at ChildProcess.emit (node:events:519:28)
   at ChildProcess._handle.onexit (node:internal/child_process:294:12)
[root@localhost trilium]#
```


#### 解决方法

1.  通过npm安装pnpm

```
# 使用系统已有npm进行安装
sudo npm install -g pnpm
```

1.  验证安装

```
pnpm -v  # 预期返回版本号（应≥7.0.0）
```

1.  安装依赖项目

```
# 进入项目目录（重要！）
cd /path/to/trilium
# 安装所有依赖（包含tsx）
pnpm install
```
注意以上步骤都完成了，运行服务器（9\. 启动服务）会出现下面程序崩溃的情况
\[nodemon\] app crashed - waiting for file changes before starting...

优先使用 依赖安装失败处理（本文文末或原文章文末）
以上方法不管用在使用下面方法
(本人解决方法是首先使用下面的办法失败，最后重新按照原文的 依赖安装失败处理，然后重新8\. 安装项目依赖，下载pnpm安装tsx）
```
[nodemon] app crashed - waiting for file changes before starting…

▸ 核心问题：应用运行时异常 ▸ 常见成因：依赖缺失/配置错误/端口冲突/权限问题分步排查流程
1. 查看完整错误日志# 手动执行启动命令获取详细报错

TRILIUM_ENV=dev tsx src/main.ts

2. 验证依赖完整性# 强制重新安装所有依赖

pnpm install --force

# 检查核心依赖

pnpm list tsx typescript @types/node

3. 端口冲突检测（默认8080）
sudo lsof -i :8080 || echo “端口可用”

# 可选：更换启动端口

TRILIUM_ENV=dev PORT=8081 tsx src/main.ts

4. 文件权限修复# 递归修复项目文件权限

sudo chown -R $(whoami):$(whoami) .

find . -type d -exec chmod 755 {} \;

find . -type f -exec chmod 644 {} \;
```

访问与管理
-----

### 10\. 访问服务
本机可能绘=会访问失败只出现登录页面，但是没有内容，可以切换设备查看ip，端口8080进入
*   本机访问：`http://127.0.0.1:8080`
*   局域网访问：`http:[设备IP]:8080`

### 11\. 创建快捷启动脚本

新建`/root/run-trilium.sh`：

```sh

#!/bin/bash

source ~/.nvm/nvm.sh

nvm use v20.17.0

cd /root/trilium

TRILIUM_ENV=dev nodemon src/main.ts

```

设置可执行权限：

```sh

chmod +x /root/run-trilium.sh

```

快速启动命令：

```sh

proot-distro login manjaro -- /root/run-trilium.sh

```

常见问题处理
------

### 网络问题解决方案

```sh

pacman -S proxychains

vim /etc/proxychains.conf  # 配置代理地址

proxychains npm install    # 通过代理执行

```

### 依赖安装失败处理

```sh

rm -rf node_modules ~/.npm/_cacache  # 清除缓存

npm cache clean --force

```

### 性能优化建议

1.  关闭手机省电模式
2.  授予Termux后台运行权限
3.  避免同时运行大型应用
4.  定期使用`pacman -Scc`清理软件包缓存

* * *

**注意事项**

1.  首次启动后建议立即设置管理员密码
2.  数据目录默认位于`/root/trilium/data`
3.  建议配合Termux:Widget创建桌面快捷方式
4.  跨设备访问时注意网络安全

通过这个优化版教程，用户可以在20分钟内完成全套部署流程。实际测试在骁龙8 Gen2设备上，服务启动时间约15秒，笔记操作响应延迟小于300ms。

**Comment by d3v19** [comment]
Wow this is exactly what I was trying to a few days ago and failed. Thanks man. Btw the dependency install went fine but nvm install was having some issues due to curl certificate. 

`curl: (77) error setting certificate file: /etc/ssl/certs/ca-certificates.crt`

Fixed that by installing a package with

```
pacman -S ca-certificates-utils
```

Edit: Now that I see @hewoyu already added this solution lol.

**Comment by jasongwq** [comment]
我打包了一个可以直接安装在termux上的压缩包，解压即可直接运行
I've packaged a compressed archive that can be directly used on Termux—just extract and run.

https://github.com/jasongwq/Trilium-termux/

**Reply by contributor** [reply]
Cool. Sounds like this could be used for offline access to trilium notes on phone? Can I also setup sync to existing instances to keep offline database up to date?
